presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/PooledHiveMetastoreClientFactory.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive.metastore.thrift;
+
+import com.google.common.cache.CacheLoader;
+import com.google.common.net.HostAndPort;
+import io.airlift.units.Duration;
+import io.prestosql.plugin.hive.util.ObjectPool;
+
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.inject.Qualifier;
+
+import java.io.Closeable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.reflect.Reflection.newProxy;
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+import static java.util.Objects.requireNonNull;
+
+public class PooledHiveMetastoreClientFactory
+        implements ThriftMetastoreClientFactory, Closeable
+{
+    private final ObjectPool<Key, ThriftMetastoreClient> pool;
+
+    @Inject
+    public PooledHiveMetastoreClientFactory(@ForPooledMetastore ThriftMetastoreClientFactory delegate)
+    {
+        this.pool = new ObjectPool<>(
+                new CacheLoader<>()
+                {
+                    @Override
+                    public ThriftMetastoreClient load(Key key)
+                            throws Exception
+                    {
+                        return delegate.create(key.getAddress(), key.getDelegationToken());
+                    }
+                },
+                1000,
+                Duration.valueOf("30s"));
+    }
+
+    @Override
+    public ThriftMetastoreClient create(HostAndPort address, Optional<String> delegationToken)
+    {
+        ObjectPool.Lease<ThriftMetastoreClient> lease = pool.get(new Key(delegationToken, address));
```

Comment:
with this implementation we can end up returning a series of broken connections to user which will make retry mechanism in the layer above non-effective.
Example case when we will return a series of broken connections is when we end up with having many connections in the queue in pool after a spike. With limited traffic after the spike some of the connections will time out due to inactivity (probably many at more-or same time).

I think when we got a failure on pooled connection we should return a brand new connection to same caller on next `create` request. I have no concrete proposal how to implement that nicely though.

Alternatively we can add some connection testing mechanism, so connection is verified before being returned to caller. Less bullet proof - but maybe good enough.